client/src/analysis/Functions.res

```diff
@@ -128,11 +129,15 @@ let builtins = (t: t): list<Function.t> => t.builtinFunctions |> List.map(~f=Fun
 let calculateAllowedFunctionsList = (props: props, t: t): list<Function.t> => {
   // We hide functions that are deprecated unless they are in use
   let filterAndSort = (fns: list<Function.t>): list<Function.t> => {
-    let isUsedOrIsNotDeprecated = (f: Function.t): bool =>
-      if f.fnDeprecated {
-        Map.get(~key=FQFnName.toString(f.fnName), props.usedFns)
-        |> Option.unwrap(~default=0)
-        |> (count => count > 0)
+    let isUsed = (f: Function.t): bool =>
+      Map.get(~key=FQFnName.toString(f.fnName), props.usedFns)
+      |> Option.unwrap(~default=0)
+      |> (count => count > 0)
```

Comment:
I found the old form much easier to read here.

This new form also does a bunch more unnecessary work.

Reply:
I think you're referring to `isUsed` being extracted out of `isUsedOrIsNotDeprecated`?


If I revert that extraction; here's what I'm left with:
```
let filterAndSort = (fns: list<Function.t>): list<Function.t> => {
  let isUsedOrIsNotDeprecated = (f: Function.t): bool =>
    if f.fnDeprecated {
      Map.get(~key=FQFnName.toString(f.fnName), props.usedFns)
      |> Option.unwrap(~default=0)
      |> (count => count > 0)
    } else {
      true
    }

  let isExperimentalAndOptedIn = (f: Function.t): bool =>
    // TUPLETODO remove this filter when the experimental setting is removed
    switch f.fnName {
    | Stdlib(fnName) if String.startsWith(~prefix="Tuple", fnName.module_) => props.allowTuples
    | _ => false
    }

  let fnNameWithoutVersion = ...

  fns
  |> List.filter(~f=(f: Function.t) => isUsedOrIsNotDeprecated(f) || isExperimentalAndOptedIn(f))
  ...
```

`isUsedOrIsNotDeprecated` is back as it was, and the new function has been adjusted.

This leaves me with one scenario covered poorly:
what should happen when a function is deprecated _and_ in-progress _and_ the user has opted in?

At this point, I'm tempted to either:
- throw an additional `if f.fnDeprecated` condition within `isExperimentalAndOptedIn`
- update the filter to ` || (!f.fnDeprecated && isExperimentalAndOptedIn(f))`

Both of these feel a big ugly to me, which is why I kept it as-is.
What do you think - do you like either of these?

What unnecessary work is done?

Reply:
I'm overthinking this. I'm just going  to go with `|| (!f.fnDeprecated && isExperimentalAndOptedIn(f))`

Reply:
The unnecessary work is calculating used for non-deprecated functions. I had thought this was a hot loop but now I'm not sure.

I think the logic is a bit twisted here. I think you want an `&&`: does it pass both the experimental check and the deprecated check. Which says to me that the checks need to be reshuffled to allow this.

Reply:
maybe `!f.fnDeprecated || isExperimentalAndOptedIn(f) || isUsed(f)` ?

Reply:
OK I think this is better:
```
let filterAndSort = (fns: list<Function.t>): list<Function.t> => {
  let isUsed = (f: Function.t): bool =>
    Map.get(~key=FQFnName.toString(f.fnName), props.usedFns)
    |> Option.unwrap(~default=0)
    |> (count => count > 0)

  let isNonExperimentalOrOptedIn = (f: Function.t): bool =>
    // TUPLETODO remove this filter when the experimental setting is removed
    switch f.fnName {
    | Stdlib(fnName) if String.startsWith(~prefix="Tuple", fnName.module_) => props.allowTuples
    | _ => true
    }

  let fnNameWithoutVersion = ...

  fns
  |> List.filter(~f=(f: Function.t) => {
    if f.fnDeprecated {
      isUsed(f)
    } else {
      isNonExperimentalOrOptedIn(f)
    }
  })
```

Reply:
no, that's still wrong.

Reply:
I think you also want to allow it if it's used but tuples are disabled?

Reply:
```
// check it passes the deprecation check and the experimental check
let allowed = (!deprecated || used) && (!experimental || optedIn || used)
```

Reply:
I think this fits well:
```
|> List.filter(~f=(f: Function.t) => {
  if f.fnDeprecated {
    isUsed(f)
  } else {
    isNonExperimentalOrOptedIn(f) || isUsed(f)
  }
})
```
teasing apart `experimental` and `optedIn` checks is a bit awkward, unless we define centrally some 'Experiment' DU. 

Maybe that'd be worth it, though I'm not sure where that would live. Personally that seems like overkill with 1 'experiment'.

Reply:
Yeah agreed.